### PR TITLE
Replace SchedulerDO with a plain D1-backed scheduler

### PR DESCRIPTION
**Pull request by ColeMurray**
## Summary

- replace the stateless `SchedulerDO` with a plain `Scheduler` module invoked directly by cron,
  automation routes, and webhook handlers
- inject automation run completion into `SessionDO` and remove the `DOFetcherAdapter` and the
  `SCHEDULER` environment binding
- port the scheduler tests to the module and add a real-D1 gate proving two concurrent triggers plus
  a tick admit exactly one run
- remove the Cloudflare `SchedulerDO` binding/export and add Terraform `deleted_classes` support with
  the documented two-phase removal path

`SchedulerDO` held no state — it was a Durable Object purely as an RPC target. Every call was a
`fetch()` into a singleton `idFromName("global-scheduler")` whose authoritative concurrency check
already lived in D1. Removing the indirection drops a DO namespace, the fetcher adapter, and the
`SCHEDULER` binding without changing the admission semantics.

## Durable Object migration

Class removal does not disable surviving bindings, so this uses a new migration tag rather than the
usual disable/re-enable dance:

- `terraform/modules/cloudflare-worker` gains a `deleted_classes` variable; when set, the migration is
  emitted *with* `enable_durable_object_bindings = true` so `SESSION` stays attached
- the production workflow stages the v2→v3 deletion only when `terraform state pull` still reports
  `migration_tag = "v2"`, so the release-specific migration never becomes a permanent default for
  fresh deployments (which start at v1 and never created `SchedulerDO`)

## Verification

- `npm run build -w @open-inspect/shared`
- `npm run typecheck` (all packages)
- `npm run lint -w @open-inspect/control-plane`
- `npx prettier --check` over the touched paths
- `terraform fmt -check -recursive terraform/`
- `npm test -w @open-inspect/control-plane` — 193 files, 2,995 tests passed
- `npm run test:integration -w @open-inspect/control-plane` — 74 files, 928 tests passed
- `npm test -w @open-inspect/slack-bot` — 34 files, 420 tests passed

## Notes

This is a port of an internal change that was authored before `BackgroundJobDispatcher` was renamed
to the `BackgroundTasks` port. The three original commits are preserved; the scheduler now takes
`BackgroundTasks` and `SessionDO` passes its existing `backgroundTasks` field, which is the only
adaptation relative to the original.

Linear: COL-51

## Summary by CodeRabbit

- **New Features**
  - Scheduler operations now run directly within the control plane for more consistent automation handling.
  - Callback delivery supports configurable per-attempt timeouts, including disabling timeouts.

- **Bug Fixes**
  - Prevented concurrent or duplicate triggers from skipping schedule slots.
  - Improved automation completion updates and stale schedule handling.
  - Deployment checks now fail safely when state cannot be read.

- **Chores**
  - Added deployment support for safely retiring legacy scheduler infrastructure.
  - Updated migration guidance and integration coverage.

### CODE REVIEW

File: .github/workflows/terraform.yml [hunk 1]
```diff
@@ -180,6 +180,14 @@ jobs:
             -backend-config='endpoints={s3="https://${{ secrets.CLOUDFLARE_ACCOUNT_ID }}.r2.cloudflarestorage.com"}'
         working-directory: ${{ env.TF_WORKING_DIR }}
 
+      - name: Stage SchedulerDO deletion migration
+        run: |
+          current_tag="$(terraform state pull | jq -r '.resources[]? | select(.module == "module.control_plane_worker" and .type == "cloudflare_worker_version" and .name == "this") | .instances[0].attributes.migration_tag // empty')"
+          if [ "$current_tag" = "v2" ]; then
+            printf '%s\n' '{"control_plane_migration_tag":"v3","control_plane_migration_old_tag":"v2","control_plane_deleted_classes":["SchedulerDO"]}' > scheduler-do-retirement.auto.tfvars.json
+          fi
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply .github/workflows/terraform.yml:194]: Done in `11605128d`. Both jobs now `set -euo pipefail` and read state into a variable before parsing it, so a failed `terraform state pull` stops the release instead of degrading into an empty tag that silently skips the deletion migration.

One thing worth recording: I verified an absent/empty state still exits 0 under `pipefail` (empty string, initialized-but-empty state JSON, and an undeployed version with no `migration_tag` all yield an empty tag cleanly). That matters because fresh deployments hit this step on every run and must not be failed by it.

File: packages/control-plane/src/scheduler/scheduler.ts [hunk 2]
```diff
@@ -232,16 +233,32 @@ export async function resolveAutomationProviderAuth(
...
+    if (request.method === "POST" && path === "/internal/tick") return this.tick();
+    if (request.method === "POST" && path === "/internal/trigger") {
+      return this.trigger(await request.json());
+    }
+    if (request.method === "POST" && path === "/internal/event") {
+      return this.event(await request.json());
+    }
+    if (request.method === "POST" && path === "/internal/run-complete") {
+      return this.runComplete(await request.json());
+    }
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/scheduler/scheduler.ts:259]: The `SyntaxError` behaviour is right, but `dispatch()` isn't reachable from untrusted input. Its only callers are `scheduler.test.ts` and the three integration-test stubs; production paths call `trigger()`, `event()` and `runComplete()` directly with already-parsed objects, and each validates with a zod `safeParse` that returns `badJsonRequest(...)` on failure. Adding try/catch here would only harden a test helper, so I'd rather leave the parse boundary where the real callers are.

File: packages/control-plane/src/session/callback-delivery.ts [hunk 3]
```diff
@@ -15,12 +15,18 @@ interface DeliveryResult {
 export async function deliverWithRetry(
   send: (signal: AbortSignal) => Promise<Response>,
   sleep: (ms: number) => Promise<void>,
-  onFailure: (failure: DeliveryFailure) => void | Promise<void>
+  onFailure: (failure: DeliveryFailure) => void | Promise<void>,
+  options: { attemptTimeoutMs?: number | null } = {}
 ): Promise<DeliveryResult> {
+  const attemptTimeoutMs =
+    options.attemptTimeoutMs === undefined ? CALLBACK_ATTEMPT_TIMEOUT_MS : options.attemptTimeoutMs;
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/session/callback-delivery.ts:22]: I think this misreads the guideline. The rule in `CLAUDE.md` ("Define each default value exactly once. Extract to a named constant") sits under *Durations and timeouts*, and the duration default here already is one — `CALLBACK_ATTEMPT_TIMEOUT_MS`, applied on the next line. The `= {}` is an empty options bag, not a default value in the sense the rule governs; naming it wouldn't define anything that isn't already named.

File: packages/control-plane/test/integration/scheduler.test.ts [hunk 4]
```diff
@@ -552,6 +557,91 @@ describe("SchedulerDO (integration)", () => {
...
+          headers: { "Content-Type": "application/json" },
+          body: JSON.stringify({ automationId: "auto-concurrent-admission" }),
+        });
+      const schedulers = [
+        getSchedulerStub(schedulerEnv),
+        getSchedulerStub(schedulerEnv),
+        getSchedulerStub(schedulerEnv),
+      ];
+
+      const results = await Promise.allSettled([
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/test/integration/scheduler.test.ts]: Confirmed and fixed in `11605128d`. You're right that the gate passed when one request launched and the other two threw — `responses.some(r => r.status === 201)` was satisfied by the single fulfilled response and every downstream assertion still held.

Now uses `Promise.all` so a rejected entry point fails the test outright, and asserts the loser outcomes explicitly: either a trigger won (`[201, 409]` and `tick.processed === 0`) or the tick won (`[409, 409]` and `tick.processed === 1`). That's the property the removed Durable Object used to provide by serializing callers, so it should be pinned down rather than tolerated.

File: terraform/modules/cloudflare-worker/main.tf [hunk 5]
```diff
@@ -102,15 +102,16 @@ resource "cloudflare_worker_version" "this" {
...
   # Durable Object migrations
-  # Phase 1 (enable_durable_object_bindings=false): Apply migrations WITHOUT bindings
-  # Phase 2 (enable_durable_object_bindings=true): Add bindings WITHOUT migrations
+  # Initial class creation uses bindings disabled for phase 1. Class deletion
+  # emits a migration with surviving bindings still enabled.
   # Note: Free plans require new_sqlite_classes instead of new_classes
   # When new_sqlite_classes is set, only those classes are declared as new (incremental migration).
   # When empty, all DO classes are declared as new (fresh deployment).
-  migrations = length(var.durable_objects) > 0 && !var.enable_durable_object_bindings ? {
+  migrations = (length(var.durable_objects) > 0 || length(var.deleted_classes) > 0) && (!var.enable_durable_object_bindings || length(var.deleted_classes) > 0) ? {
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply terraform/modules/cloudflare-worker/main.tf:110]: Good catch — fixed in `11605128d`. Confirmed reachable: with `deleted_classes` non-empty and `enable_durable_object_bindings = false`, `local.bindings` drops every DO binding while the migration still emits, and the CI step sets `deleted_classes` without touching that flag, so an existing v2 deployment still in phase 1 would hit it.

Added a `lifecycle.precondition` on `cloudflare_worker_version` requiring `enable_durable_object_bindings = true` whenever `deleted_classes` is non-empty, so it fails at plan time as you suggested. `terraform validate` and `terraform fmt -check -recursive` both pass.

File: packages/control-plane/src/session/durable-object.ts [hunk 6]
```diff
@@ -436,18 +436,17 @@ export class SessionDO extends DurableObject<Env> {
    */
   private get callbackService(): CallbackNotificationService {
     if (!this._callbackService) {
-      // Wrap SchedulerDO namespace as a Fetcher for automation callbacks
-      const schedulerCallback = this.env.SCHEDULER
-        ? new DOFetcherAdapter(this.env.SCHEDULER, "global-scheduler")
+      const scheduler = this.db
+        ? new Scheduler(this.db, this.env, this.backgroundTasks)
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/session/durable-object.ts:440]: Fair observation, and the numbers are right (1,989 / 1,552 lines, and `Scheduler` does import seven things from `session/`). But the coupling isn't introduced here — `SessionDO` already reached the scheduler through the `SCHEDULER` binding and `DOFetcherAdapter`; this PR changes the mechanism, not the direction. The `Response`-returning callback is a real smell, agreed.

Extracting a run-completion application service with a typed result and moving `Response` conversion to the route layer is the right shape, but it's a design change well beyond retiring the Durable Object, and it'd sit on top of an already-deployed code path. I'd rather land the removal and do that extraction as a focused follow-up. Happy to file it.

File: packages/control-plane/src/db/automation-store.ts [hunk 7]
```diff
@@ -934,9 +941,15 @@ export class AutomationStore {
         this.db
           .prepare(
             `UPDATE automations SET next_run_at = ?, updated_at = ?
-             WHERE id = ? AND deleted_at IS NULL`
+             WHERE id = ? AND deleted_at IS NULL
+               AND (next_run_at IS NULL OR next_run_at < ?)`
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/db/automation-store.ts]: Agreed, and fixed in `11605128d`. Both advance sites now compare-and-set on the claimed slot (`WHERE next_run_at = ?` bound to `fromSlot`) rather than accepting any larger timestamp, via a new `ScheduleAdvance { fromSlot, nextRunAt }`. `params.scheduledAt` is already the slot the tick read (`scheduledAt: automation.next_run_at`), so the caller has the right value to hand over.

Your scenario reproduces: cron `* * * * *`, slot 12:00 — tick A at 12:00:59 advances to 12:01, then tick B at 12:01:01 advances to 12:02 because `12:01 < 12:02` passes, and 12:01 never fires. Added a real-D1 regression test for exactly that handover.

Worth flagging: the existing `insertSkippedInvocation` integration test *asserted* the double advance (first skip 1000→2000, duplicate skip then 2000→3000, expecting 3000). I changed it to expect 2000. The re-collision it was guarding against can't occur, because the winning firing already moved `next_run_at` off the claimed slot. Left `advanceNextRunAt` alone — it's pre-existing and untouched by this PR.

File: packages/control-plane/src/db/automation-store.ts [hunk 8]
```diff
@@ -885,9 +886,15 @@ export class AutomationStore {
...
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/db/automation-store.ts]: Partly pushing back. On the specific claim: the `next_run_at IS NULL` branch re-arming a paused automation is pre-existing, not introduced here — before this PR the advance was deliberately unconditional (`WHERE id = ? AND deleted_at IS NULL`), so it advanced over `NULL` too. This PR narrowed that, and the slot-ownership change on the sibling thread narrows it further to the claimed slot, which also drops the `NULL` branch entirely.

The broader ask — a database-owned CAS over `enabled`, deletion, trigger type and claimed slot in the same transaction as invocation creation, with a single `admitted | overlap | duplicate | stale` result — is a redesign of the admission protocol rather than a fix to this change. Worth doing, but as its own PR where the result type can replace the repair branches properly.

File: .github/workflows/terraform.yml [hunk 9]
```diff
@@ -341,6 +349,14 @@ jobs:
             -backend-config='endpoints={s3="https://${{ secrets.CLOUDFLARE_ACCOUNT_ID }}.r2.cloudflarestorage.com"}'
         working-directory: ${{ env.TF_WORKING_DIR }}
 
+      - name: Stage SchedulerDO deletion migration
+        run: |
+          current_tag="$(terraform state pull | jq -r '.resources[]? | select(.module == "module.control_plane_worker" and .type == "cloudflare_worker_version" and .name == "this") | .instances[0].attributes.migration_tag // empty')"
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply .github/workflows/terraform.yml]: The mechanism checks out — I confirmed against the provider schema that `migration_tag` is computed and documented as *"Set when the version is deployed. Omitted if the version has not been deployed"*. So a run that creates the version and dies before `cloudflare_workers_deployment` does leave an undeployed version with no tag.

Where I differ is the consequence. It self-heals: a version that deploys without the migration leaves the worker still at v2, state reports v2 again, and the next run re-stages the deletion. Combined with the fail-closed read on the sibling thread, the remaining window is one failed apply rather than a permanently-skipped migration. Deriving the decision from the deployed version via the API or persisting a durable marker is real machinery to add to a one-shot, self-retiring path that has already executed successfully. I'd rather leave it and delete the whole step once no deployment reports v2.
